sc_cost_meter/utils.py

```diff
@@ -0,0 +1,208 @@
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+os.environ["AWSPRICING_USE_CACHE"] = "1"  # enable cache for awspricing
+
+def get_ec2_client():
+  return boto3.client('ec2')
+
+def get_ssm_client():
+  return boto3.client('ssm')
+
+def get_ce_client():
+  return boto3.client('ce')
+
+def get_meteringmarketplace_client():
+  return boto3.client('meteringmarketplace')
+
+def get_instances(states):
+  '''
+  Get all EC2 instances in the specfied states
+  :param states: list of states (['pending'|'running'|'shutting-down'|'terminated'|'stopping'|'stopped'])
+         https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceState.html
+  :return: list of instances in the state specified by states, empty list if none are
+           found in the specified states
+  '''
+  client = get_ec2_client()
+  response = client.describe_instances(
+    Filters=[
+       {
+          'Name': 'instance-state-name',
+          'Values': states
+       }
+    ]
+  )
+
+  instances = []
+  reservations = response["Reservations"]
+  if reservations:
+    for reservation in reservations:
+        for instance in reservation["Instances"]:
+          instances.append(instance)
+
+  return instances
+
+def get_ec2_hourly_on_demand_pricing(instance):
+  '''
+  Get the hourly on demand pricing for an instance
+  :param instance: an EC2 in the AWS account
+  :return: the hourly on demand price of the instance
+          https://aws.amazon.com/ec2/pricing/on-demand/
+  '''
+  instance_region = instance['Placement']['AvailabilityZone'][0:-1]
+  instance_id = instance['InstanceId']
+  instance_type = instance['InstanceType']
+
+  client = get_ssm_client()
+  response = client.describe_instance_information(
+    Filters=[
+       {
+          'Key': 'InstanceIds',
+          'Values': [instance_id]
+       }
+    ]
+  )
+  ssm_instance_info = response['InstanceInformationList'][0]
+  instance_platform = ssm_instance_info['PlatformType']
+
+  ec2_offer = awspricing.offer('AmazonEC2')
+  price = ec2_offer.ondemand_hourly(
+    instance_type,
+    operating_system=instance_platform,
+    region=instance_region
+  )
+
+  log.debug(f'price for {instance_id} is {price}')
+  return price
+
+def get_ec2_cost(tags):
+  '''
+  Get the total cost of all EC2s with an attached set of tags.
+  :param tags: a dictionary of tags (i.e. {"Key": "customerId", "Values": ["1234ABCD"]})
+  :return: the total cost of all EC2s
+          Note: AWS billing history shows that cost data is only updated every 24 hours
+          therefore the response from get_cost_and_usage API may lag up to one day.
+          This means that the only way to get consistent billing info is to query the
+          previous day's billing.
+  '''
+  client = get_ce_client()
+
+  # Get the cost from the previous day
+  current_time = datetime.utcnow()
+  start_time = current_time - timedelta(days=2)
+  end_time = current_time - timedelta(days=1)
+
+  result = client.get_cost_and_usage(
```

Comment:
Why did you decide to use this API (which only returns EC2 information) rather than using the Cost and Usage reports ( https://docs.aws.amazon.com/cur/latest/userguide/what-is-cur.html) which I believe returns comprehensive information about the account's bill?

Reply:
to clarify, I assume you mean the cost and usage report APIs here? https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cur.html

at first glance i'm not sure how that's better than using the cost explorer APIs.  it seems like the cost and usage API is used to create reports, in some format, which are deposited into an S3 bucket.  I'm guessing you do the extra step of retrieving the file and data after generating the report.  it seems like you can achieve the same result by using the cost explorer API except that the returned value from the cost explorer API is the cost data?  I think that's how it works but will investigate deeper.

Reply:
> to clarify, I assume you mean the cost and usage report APIs here? https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cur.html

Good sleuthing, I did not find the boto client for this service.  Yes, I expect that we would call [put_report_definition()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cur.html#CostandUsageReportService.Client.put_report_definition)  once to set up the system and that the generated report (updated once a day I think) would be read by this Lambda to decide how much to bill each user.



>  it seems like the cost and usage API is used to create reports, in some format, which are deposited into an S3 bucket.

Yes, I believe that's correct.

>  I'm guessing you do the extra step of retrieving the file and data after generating the report. it seems like you can achieve the same result by using the cost explorer API except that the returned value from the cost explorer API is the cost data? I think that's how it works but will investigate deeper.


> at first glance i'm not sure how that's better than using the cost explorer APIs.

The improvement would be getting all cost data rather than just EC2 cost data.  For example there should be S3 cost data in the report.

Reply:
I think using the cost explorer API is the better option.  My refactor now gets costs for all resources that are tagged with the marketplace customer ID.